### PR TITLE
Processing data

### PR DESCRIPTION
**Pull request by lengushuru**
### Hello CR 👋 

**In this PR The following changes are suggested as per [these ](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/Processing_data_model_project_v1_1.md)specifications:**

- Create model classes for all entities as shown in [ERD ](https://github.com/Hisoka37/Rails_Blog/images/blog_app_erd_v1_1.png).
- Set up associations between models.
- Create custom methods
- Users
- 
    1.   A method that returns the 3 most recent posts for a given user.
    
- Posts

      1.  A method that updates the posts counter for a user.
      3.  A method that returns the 5 most recent comments for a given post.

- Comments
      1.  A method that updates the comments counter for a post.

- Likes

       1.  A method that updates the likes counter for a post.

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -1,2 +1,15 @@
 class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
```
Comment (rloterh) [inline app/models/post.rb]: - [ ] The `id` of _Post_ is supposed to reference the foreign key `author_id` of the _User_ model (and also declared in the _User_ model), and hence it's redundant to declare `author_id` as a foreign key in _Post_ model. Let's simply do something like:
```ruby
# app/models/post.rb
belongs_to :author, class_name: 'User'

File: app/models/comment.rb [hunk 2]
```diff
@@ -1,2 +1,10 @@
 class Comment < ApplicationRecord
+  belongs_to :post
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
```
Comment (rloterh) [inline app/models/comment.rb]: - [ ] Again, the `id` of _Comment_ is supposed to reference the foreign key `author_id` of the _User_ model (and also declared in the _User_ model), and hence it's redundant to declare `author_id` as a foreign key in _Comment_ model. Let's simply do something like:
```ruby
# app/models/;comment.rb
belongs_to :author, class_name: 'User'

File: app/models/like.rb [hunk 3]
```diff
@@ -1,2 +1,10 @@
 class Like < ApplicationRecord
+  belongs_to :post, class_name: 'Post', foreign_key: :post_id
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
```
Comment (rloterh) [inline app/models/like.rb]: - [ ] Again, the `id` of _Like_ is supposed to reference the foreign key `author_id` of the _User_ model (and also declared in the _User_ model), and hence it's redundant to declare `author_id` as a foreign key in _Like_ model. Let's simply do something like this:
```ruby
# app/models/;like.rb
belongs_to :author, class_name: 'User'

File: app/models/post.rb [hunk 4]
```diff
@@ -1,2 +1,15 @@
...
+  has_many :comments, class_name: 'Comment', foreign_key: :post_id
+  has_many :likes, class_name: 'Like', foreign_key: :post_id
+
+  after_create :update_user_posts_counter
+
+  def update_user_posts_counter
```
Comment (rloterh) [inline app/models/post.rb:8]: - [ ] OPTIONAL
Following best practices `update_user_posts_counter` should be marked as _private_ after implementing the `after_create` method.

File: app/models/comment.rb [hunk 5]
```diff
@@ -1,2 +1,10 @@
...
+
+  attribute :text, :text
+  after_create :update_post_comments_counter
+  def update_post_comments_counter
+    post.increment!(:comments_counter)
```
Comment (rloterh) [inline app/models/comment.rb:8]: - [ ] OPTIONAL
Following best practices `update_comments_counter` should be marked as _private_ after implementing the `after_create` method.

File: app/models/like.rb [hunk 6]
```diff
@@ -1,2 +1,10 @@
...
+
+  after_create :update_post_likes_counter
+
+  def update_post_likes_counter
+    post.increment!(:likes_counter)
+  end
```
Comment (rloterh) [inline app/models/like.rb:9]: - [ ] OPTIONAL
Following best practices `update_likes_counter` should be marked as _private_ after implementing the `after_save` method.

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Lengushuru, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 
🏁 🚦  🚗 _ 🏃 🌴  🏠 🌳 🏘️ 🏡  

## Highlights 👍  
- Good job so far with creating model associations!
- ReadMe looks great!

## Required Changes ♻️
_Kindly check the inline comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

_Kindly consider my inline comments marked **[OPTIONAL]** under this review_

**You can also consider:**
Let's improve the look of the ReadMe by removing the underlined:

image

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 1 above)
Comment (lengushuru) [reply app/models/post.rb]: Thank you @rloterh, I will make those changes 💯

File: app/models/post.rb [hunk 7]
```diff
@@ -1,2 +1,15 @@
 class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  has_many :comments, class_name: 'Comment', foreign_key: :post_id
+  has_many :likes, class_name: 'Like', foreign_key: :post_id
+
+  after_create :update_user_posts_counter
+
+  def update_user_posts_counter
```
Comment (tufoinnkuo10) [inline app/models/post.rb:8]: - [optional] please, like the previous reviewer noted, kindly consider using one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) - after_save to call the #update methods then make the method private . Nice work 😄

**Review by tufoinnkuo10 — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi @lengushuru  

Your project is complete! Great job for implementing the changes requested by the previous reviewer 🥇. There is nothing else to say other than... it's time to merge it 🍾🚢 :shipit: . Congratulations!  🎉💯🌟 

## Good Points 👍
- Descriptive Pull Request ✔️
- Association added ✔️
- Methods ok ✔️
- No linter errors ✔️ 

<br>
Cheers..Happy coding!..and keep soaring higher! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 2 more limited reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**